src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,149 @@
+        if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+            size++;
+            return;
+        }
+        Node<T> currentNode = getNode(index);
+        newNode.next = currentNode;
+        newNode.prev = currentNode.prev;
+        currentNode.prev = newNode;
+        currentNode.prev.prev.next = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIfIndexOutOfBounds(index);
+        Node<T> currentNode = getNode(index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIfIndexOutOfBounds(index);
+        Node<T> currentNode = getNode(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        return unlink(getNode(index), index);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value == object
+                    || currentNode.value != null && currentNode.value.equals(object)) {
+                remove(i);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == 0;
+    }
+
+    public T unlink(Node node, int index) {
+        checkIfIndexOutOfBounds(index);
+        Node<T> currentNode;
+        if (index == 0) {
+            currentNode = head;
+            head = currentNode.next;
+            size--;
+            return currentNode.value;
+        }
+        if (index == size - 1) {
+            currentNode = tail;
+            tail = currentNode.prev;
+            tail.next = null;
+            currentNode.prev = null;
+            size--;
+            return currentNode.value;
+        }
+        currentNode = getNode(index);
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+        node.prev = null;
+        node.next = null;
+        size--;
+        return currentNode.value;
+    }
+
+    private void checkIfIndexOutOfBounds(int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Invalid index: "
+                    + index
+                    + "size , "
+                    + size);
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> node = head;
+        for (int i = 0; i < index; i++) {
+            node = node.next;
+        }
+        return node;
```

Comment:
checklist issue - iterate from the side which is closer
https://github.com/mate-academy/jv-linked-list/blob/master/checklist.md#dont-iterate-your-nodes-from-the-head-to-the-tail-or-vice-versa